File: 5. Longest Palindromic Substring/expandCenter.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 0:
            return ""
        start, end = 0, 0
        for i in range(len(s)):
            len1 = len(self.expandCenter(s, i, i))
            len2 = len(self.expandCenter(s, i, i + 1))
            length = max(len1, len2)
            if length > end - start:
                start = i - (length - 1) // 2
                end = i + length // 2
        return s[start:end + 1]
    
    def expandCenter(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1:right]

File: 5. Longest Palindromic Substring/expandCenter.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size() == 0) return "";
        int start = 0, end = 0;
        for (int i = 0; i < s.size(); i++) {
            int len1 = expandCenter(s, i, i);
            int len2 = expandCenter(s, i, i + 1);
            int len = max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substr(start, end - start + 1);
    }

    int expandCenter(string s, int left, int right) {
        while (left >= 0 && right < s.size() && s[left] == s[right]) {
            left -= 1;
            right += 1;
        }
        return right - left - 1;
    }
};